### PR TITLE
Bugfix - definition name index skipping with D/I

### PR DESCRIPTION
**Pull request by adkian-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

   - bug fix                          

#### API Impact

Extends Builder `build` by adding an optional `forceModName` parameter 
#### Backend Code Generation Impact

Definitions created with the D/I API that have the same name will be indexed as `Defn`, `Defn_1`, ... instead of `Defn`, `Defn_2`, ... . This matches the behavior for module instances created without D/I (that is, as `Module(new Defn())`

#### Desired Merge Strategy

 - Squash: The PR will be squashed and merged (choose this if you have no preference.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.3.x`, [small] API extension: `3.4.x`, API modification or big change: `3.5.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Review by mwachs5 — COMMENTED** [review]
Can you add a few more tests:

1. Where you create a `Module` and a `Definition/Instance` of the same thing in the same scope -- should still end up with an `AddOne` and an `AddOne_1`.
2. Can you do more than 2 things, like 3? to show there is no other weird skipping

**Comment by adkian-sifive** [comment]
Adding these tests and a test for naming nested submodules as per @jackkoenig 's review

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 1]
```diff
@@ -740,14 +740,15 @@ private[chisel3] object Builder extends LazyLogging {
...
+  private[chisel3] def build[T <: BaseModule](f: => T, dynamicContext: DynamicContext, forceModName: Boolean = true): (Circuit, T) = {
     dynamicContextVar.withValue(Some(dynamicContext)) {
       ViewParent // Must initialize the singleton in a Builder context or weird things can happen
                  // in tiny designs/testcases that never access anything in chisel3.internal
       checkScalaVersion()
       logger.info("Elaborating design...")
       val mod = f
-      mod.forceName(None, mod.name, globalNamespace)
+      if (forceModName) // This avoids definition name index skipping with D/I
+        mod.forceName(None, mod.name, globalNamespace)
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:751]:
```suggestion
      if (forceModName) { // This avoids definition name index skipping with D/I
        mod.forceName(None, mod.name, globalNamespace)
      }
```

File: src/test/scala/chiselTests/experimental/hierarchy/Examples.scala [hunk 2]
```diff
@@ -45,7 +58,7 @@ object Examples {
     i0.in := in
     i1.in := i0.out
     out := i1.out
-  }
+  }
```
Comment (adkian-sifive) [inline src/test/scala/chiselTests/experimental/hierarchy/Examples.scala:61]:
```suggestion
  }
```

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 3]
```diff
@@ -19,7 +19,7 @@ import logger.LazyLogging
 import scala.collection.mutable
 
 private[chisel3] class Namespace(keywords: Set[String]) {
-  private val names = collection.mutable.HashMap[String, Long]()
+  private[chisel3] val names = collection.mutable.HashMap[String, Long]()
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:22]: Does this need to be made package private? I don't see any uses of it

Comment (adkian-sifive) [reply core/src/main/scala/chisel3/internal/Builder.scala]: Not anymore, fixed

**Review by jackkoenig — APPROVED** [review]
LGTM!

**Pull request merged**